### PR TITLE
UCL-537 DataGrid nested rows

### PR DESCRIPTION
**Pull request by szmejk**

### CODE REVIEW

File: src/components/DataGrid/DataGridBody/DataGridRow/DataGridRow.module.scss [hunk 1]
```diff
@@ -22,28 +23,120 @@
...
+  }
+
+  &.line {
+    &::before {
+      position: absolute;
+      content: "";
+      left: 50%;
+      height: 50%;
+      width: 1px;
+      background-color: var(--light-grey-border);
```
Comment (MichalGorskiOnegini) [inline src/components/DataGrid/DataGridBody/DataGridRow/DataGridRow.module.scss:129]: as discussed, the color should (most likley) be sligthly different :D

File: src/components/DataGrid/DataGridBody/DataGridRow/DataGridRow.tsx [hunk 2]
```diff
@@ -23,12 +23,20 @@ import { DataGridCell } from "../DataGridCell/DataGridCell";
...
+  item?: T;
+  indentationLevel?: number;
+  indentationLevels?: { level: number; isLastChild: boolean }[];
+  rowTemplate?: ({ item, index }: { item: T; index: number }) => ReactElement;
   headers?: HeaderCell[];
   isLoading?: boolean;
   spacing?: React.CSSProperties;
   searchValue?: string;
   disableContextMenuColumn?: boolean;
+  enableNestedRow?: boolean;
```
Comment (MichalGorskiOnegini) [inline src/components/DataGrid/DataGridBody/DataGridRow/DataGridRow.tsx:36]: shouldn't we combine nested rows related properties to a single object?

File: src/components/DataGrid/DataGridBody/DataGridRow/DataGridRow.tsx [hunk 3]
```diff
@@ -60,14 +75,129 @@ const DataGridRowComponent: ForwardRefRenderFunction<HTMLTableRowElement, Props>
...
+        <div
+          data-testid="dataGridRowConnector"
+          className={`${classes["connector"]} ${classes["vertical"]}  ${offsetLeftClass}`}
+        />
+      );
+    }
+    return null;
+  };
+
+  const renderRecurrentRow = () => {
```
Comment (MichalGorskiOnegini) [inline src/components/DataGrid/DataGridBody/DataGridRow/DataGridRow.tsx:123]: Is `recurrent` what you really meant?

File: src/components/DataGrid/DataGridBody/DataGridRow/DataGridRow.tsx [hunk 4]
```diff
@@ -60,14 +75,129 @@ const DataGridRowComponent: ForwardRefRenderFunction<HTMLTableRowElement, Props>
...
+        aria-expanded={isRowExpanded}
+        onClick={() => setIsRowExpanded(!isRowExpanded)}
+      >
+        <Icon size="0.75rem" icon={isRowExpanded ? Icons.ChevronUp : Icons.ChevronDown} />
+      </IconButton>
+    ) : null;
+
   const visibleCells = React.Children.map(children as React.ReactElement[], (child, index) => {
+    const hasNestedChildren = item && nestedItemsKey && item[nestedItemsKey];
+    const nestedChildOffset = !hasNestedChildren ? 46 : 0;
```
Comment (MichalGorskiOnegini) [inline src/components/DataGrid/DataGridBody/DataGridRow/DataGridRow.tsx:166]: The cognitive load here is quite high. The minimum solution I could suggest is to extract those 'random' numbers to some constant that gives at least minimal explanation on what it is. If you introduce a hook as stated before, you could also extract it to some function e.g `calculateIndentation`.

File: src/components/DataGrid/DataGrid.test.tsx [hunk 5]
```diff
@@ -709,3 +709,135 @@ describe("DataGrid with search", () => {
...
+  };
+};
+
+describe("DataGrid with nested rows", () => {
+  it("should render offset nested rows with connectors", async () => {
+    const { getByTitle, getByText } = createDataGridWithNestedRows();
+
+    await userEvent.click(getByTitle("Expand row"));
+
+    expect(getByText("Company 3")).toBeInTheDocument();
```
Comment (MichalGorskiOnegini) [inline src/components/DataGrid/DataGrid.test.tsx:818]: You should also check if its not available before you clicked on the expand row button.

File: src/components/DataGrid/DataGrid.test.tsx [hunk 6]
```diff
@@ -709,3 +709,135 @@ describe("DataGrid with search", () => {
...
+  const queries = render(<DataGridWithNestedRows />);
+  const dataGrid = queries.getByTestId("dataGrid");
+
+  return {
+    ...queries,
+    dataGrid
+  };
+};
+
+describe("DataGrid with nested rows", () => {
```
Comment (MichalGorskiOnegini) [inline src/components/DataGrid/DataGrid.test.tsx:812]: Perhaps we should move it to a separate testing file? Not pushing for that but I'm worried how quickly this file grows...

File: stories/DataGrid/DataGrid.stories.tsx [hunk 7]
```diff
@@ -619,6 +619,128 @@ DataGridWithSearch.args = {
...
+  return (
+    <div style={{ padding: "1rem", boxShadow: "0px 1px 5px 0px #01053214" }}>
+      <div style={{ borderRadius: ".5rem", backgroundColor: "#FFF" }}>
+        <DataGridComponent
+          {...args}
+          data={args.data}
+          nestedRowConfig={{ nestedItemsKey: "nestedItems" }}
+        >
+          {({ item }: { item: DataGridItem }) => (
+            <DataGridRow key={item.id}>
```
Comment (MichalGorskiOnegini) [inline stories/DataGrid/DataGrid.stories.tsx:632]: I'm getting lost here a bit. So to enable nested rows you need to pass the key under which you can find a nested element. That is fine, however, in DataGridRow we have enableNestedRow property that can be used as well. Is it by design, so you can turn it off for a specific row?

**Review by MichalGorskiOnegini — CHANGES_REQUESTED** [review]

File: src/components/DataGrid/DataGridHeader/DataGridHeader.tsx [hunk 8]
```diff
@@ -107,6 +109,8 @@ const DataGridHeaderComponent: ForwardRefRenderFunction<HTMLTableSectionElement,
   return (
     <thead {...rest} ref={ref}>
       <tr className={classes["row"]}>
+        {/* <td style={{ paddingRight: spacing?.paddingRight, width: "1px" }}></td>
```
[the remark this answers is not in the document]
Comment (szmejk) [reply src/components/DataGrid/DataGridHeader/DataGridHeader.tsx]: removed

File: stories/DataGrid/DataGrid.stories.tsx [hunk 9]
```diff
@@ -619,6 +619,128 @@ DataGridWithSearch.args = {
...
+          name: "Company 9",
+          id: "9",
+          type: "Stock",
+          description: "Lorem ipsum dolor sit amet"
+        }
+      ]
+    },
+    {
+      name: "Company 3",
+      id: "3",
```
[the remark this answers is not in the document]
Comment (szmejk) [reply stories/DataGrid/DataGrid.stories.tsx]: that warning was actually unrelated to the item id, but rather had to do with me not passing any key to the connector components

File: src/components/DataGrid/DataGridBody/DataGridRow/DataGridRow.tsx [hunk 10]
```diff
@@ -60,14 +75,129 @@ const DataGridRowComponent: ForwardRefRenderFunction<HTMLTableRowElement, Props>
...
+    ? classes[`border-${indentationLevel}`]
+    : classes[`border`];
+
+  className && classNames.push(className);
+  enableExpandableRow
+    ? !isRowExpanded && classNames.push(classes["border-drawer"])
+    : classNames.push(rowBorderClass);
+  isLoading && classNames.push(classes["loading"]);
+
+  const renderNestedRowConnectors = () => {
```
[the remark this answers is not in the document]
Comment (szmejk) [reply src/components/DataGrid/DataGridBody/DataGridRow/DataGridRow.tsx]: done

File: src/components/DataGrid/DataGridBody/DataGridRow/DataGridRow.module.scss (hunk 1 above)
Comment (szmejk) [reply src/components/DataGrid/DataGridBody/DataGridRow/DataGridRow.module.scss]: done

File: src/components/DataGrid/DataGridBody/DataGridRow/DataGridRow.tsx (hunk 2 above)
Comment (szmejk) [reply src/components/DataGrid/DataGridBody/DataGridRow/DataGridRow.tsx]: done

File: src/components/DataGrid/DataGridBody/DataGridRow/DataGridRow.tsx (hunk 3 above)
Comment (szmejk) [reply src/components/DataGrid/DataGridBody/DataGridRow/DataGridRow.tsx]: changed to `nested`

File: src/components/DataGrid/DataGridBody/DataGridRow/DataGridRow.tsx (hunk 4 above)
Comment (szmejk) [reply src/components/DataGrid/DataGridBody/DataGridRow/DataGridRow.tsx]: done

File: src/components/DataGrid/DataGrid.test.tsx (hunk 5 above)
Comment (szmejk) [reply src/components/DataGrid/DataGrid.test.tsx]: done

File: stories/DataGrid/DataGrid.stories.tsx (hunk 7 above)
Comment (szmejk) [reply stories/DataGrid/DataGrid.stories.tsx:632]: the `enableNestedRow` is just an internal prop and it's based on the key

**Review by MichalGorskiOnegini — APPROVED** [review]

**Pull request merged**